### ISSUE TITLE
Stateful functions

### ISSUE THREAD

**Issue by jonthegeek** [jonthegeek/factory#34]
This package completely kills the "stateful functions" use case of function factories. It wouldn't be hard to fix that, but I should give it some thought.

Two things:

* Probably build in some "standard" states (like "how many times have I been called?"). Work that out.
* When the user prints the function, also show the state information. Maybe via a print method? Can functions have classes?

**Comment by TylerGrantSmith** [comment]
Yes they can have classes.  For example, see `rlang::as_function` re: lambda functions

**Comment by TylerGrantSmith** [comment]
Could also point out that the simple `new_counter`  from 10.2.4 does not involve any arguments to the factory (contrary to the current requirements in `build_factory`)

**Comment by jonthegeek** [comment]
> Could also point out that the simple `new_counter` from 10.2.4 does not involve any arguments to the factory (contrary to the current requirements in `build_factory`)

I was gonna disagree with that being realistic... but I can totally imagine wanting something similar in a real usecase (maybe return the ith row of a specific dataframe, where i is the number of time it's been called). I'll think about what that means for the rest of the package. I can *probably* make that rule looser (and just skip all the fanciness if you don't have arguments)...

**Comment by TylerGrantSmith** [comment]
I'll submit a PR with an option in a few minutes
